src/Build.UnitTests/Graph/ParallelWorkSet_Tests.cs

```diff
@@ -0,0 +1,253 @@
+        }
+
+        private struct WorkItem
+        {
+            internal string Key { get; set; }
+
+            internal Func<string> WorkFunc { get; set; }
+        }
+
+        private ParallelWorkSet<string, string> _workSet;
+
+        [Fact]
+        public void GivenExceptionsOnCompletionThread_CompletesAndThrowsException()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = 0,
+                WorkItemsToAdd = new List<WorkItem>
+                {
+                    new WorkItem
+                    {
+                        Key = "fooKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "barKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "bazKey",
+                        WorkFunc = () => throw new Exception()
+                    }
+                },
+                ShouldExpectException = true
+            });
+        }
+
+        [Fact]
+        public void GivenExceptionsOnWorkerThread_CompletesAndThrowsExceptions()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = Environment.ProcessorCount,
+                WorkItemsToAdd = new List<WorkItem>
+                {
+                    new WorkItem
+                    {
+                        Key = "fooKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "barKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "bazKey",
+                        WorkFunc = () => throw new Exception()
+                    }
+                },
+                ShouldExpectException = true
+            });
+        }
+
+        [Fact]
+        public void GivenNoWorkItemAndMultipleWorkers_Completes()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = Environment.ProcessorCount
+            });
+        }
+
+        [Fact]
+        public void GivenNoWorkItemAndNoWorkers_Completes()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase());
+        }
+
+        [Fact]
+        public void GivenRecursiveWorkItemsAndMultipleWorkers_Completes()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = Environment.ProcessorCount,
+                WorkItemsToAdd = new List<WorkItem>
+                {
+                    new WorkItem
```

Comment:
Should the work items be generated programmatically such that the nesting level is higher than the DegreeOfParallelism, to catch any potential starvation issues?